crypto/thread/api.c

```diff
@@ -0,0 +1,265 @@
+    return 1;
+}
+
+int CRYPTO_THREAD_disable(OPENSSL_CTX *ctx)
+{
+    OPENSSL_CTX_THREADS tdata;
+    CRYPTO_WORKER worker;
+    struct list *iter, *tmp;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    tdata->threads.cap = 0;
+    tdata->enabled = 0;
+
+    list_for_each_safe(iter, tmp, &tdata->workers.available) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    list_for_each_safe(iter, tmp, &tdata->workers.busy) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    list_for_each_safe(iter, tmp, &tdata->threads.active) {
+        worker = container_of(iter, struct crypto_worker_st, list);
+        CRYPTO_THREAD_native_terminate(worker->handle);
+        CRYPTO_THREAD_native_clean(worker->handle);
+        list_del(iter);
+        OPENSSL_free(worker);
+    }
+    CRYPTO_MUTEX_unlock(tdata->lock);
+    CRYPTO_THREAD_clean(ctx, NULL);
+    return 1;
+}
+
+int CRYPTO_THREAD_cap(OPENSSL_CTX *ctx, int max_threads)
+{
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    if (tdata->enabled == 0)
+        return 0;
+
+    tdata->threads.cap = max_threads;
+    return 1;
+}
+
+size_t CRYPTO_THREAD_get_available_threads(OPENSSL_CTX *ctx)
+{
+    OPENSSL_CTX_THREADS t;
+
+    t = OPENSSL_CTX_THREADS(ctx);
+    if (t == NULL)
+        return 0;
+
+    if (t->enabled == 0)
+        return 0;
+
+    if (t->threads.cap < 0)
+        return -1;
+
+    return t->threads.cap - list_size(&t->threads.active) +
+        list_size(&t->workers.available);
+}
+
+static int CRYPTO_THREAD_spawn_worker_task(OPENSSL_CTX *ctx,
+                                           CRYPTO_WORKER_CALLBACK cb,
+                                           void *vtask)
+{
+    int available_threads;
+    CRYPTO_WORKER worker;
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    if (tdata->enabled == 0)
+        return 0;
+
+    worker = CRYPTO_WORKER_new(ctx, cb, vtask);
+    if (worker == NULL)
+        goto fail;
+
+    if (vtask != NULL)
+        list_add_tail(&worker->list, &tdata->threads.active);
+
+    available_threads = CRYPTO_THREAD_get_available_threads(ctx);
+    worker->handle = CRYPTO_THREAD_native_start(worker_main, (void*)worker, 0);
+    if (worker->handle == NULL) {
+        list_del(&worker->list);
+        goto fail;
+    }
+
+    while (vtask == NULL &&
+           CRYPTO_THREAD_get_available_threads(ctx) == available_threads)
+        ossl_sleep(500);
```

Comment:
I wanted to avoid another conditional variable broadcast from `worker_main` that the worker is ready and spawned (and only for those workers that don't start with a task already allocated). Given that the worker spawns nearly immediately after a `crypto_thread_native_start` is made, it shouldn't sleep there for a long time. As such, thought it better not to complicate the code. Do you want me to rewrite that bit?

Reply:
I must admit to being quite unhappy with seeing ossl_sleep used like this. Perhaps others have a different view point, but I would personally rewrite it.

Reply:
Maybe add callbacks to handle the sleep/wait stuff?  I also don't like seeing a sleep in openssl code.  It must be documented if it exists.

Reply:
Sleep removed.